precompiles/batch/src/tests.rs

```diff
@@ -0,0 +1,782 @@
+			.expect_log(LogsBuilder::new(Bob.into()).log1(H256::repeat_byte(0x11), vec![]))
+			.expect_log(log_subcall_succeeded(Precompile, 0))
+			.expect_log(LogsBuilder::new(Charlie.into()).log1(H256::repeat_byte(0x22), vec![]))
+			.expect_log(log_subcall_succeeded(Precompile, 1))
+			.execute_returns(Vec::new())
+	})
+}
+
+#[test]
+fn batch_all_returns() {
+	ExtBuilder::default().build().execute_with(|| {
+		batch_returns(&precompiles(), Action::BatchAll)
+			.expect_log(LogsBuilder::new(Bob.into()).log1(H256::repeat_byte(0x11), vec![]))
+			.expect_log(log_subcall_succeeded(Precompile, 0))
+			.expect_log(LogsBuilder::new(Charlie.into()).log1(H256::repeat_byte(0x22), vec![]))
+			.expect_log(log_subcall_succeeded(Precompile, 1))
+			.execute_returns(Vec::new())
+	})
+}
+
+fn batch_out_of_gas(
+	precompiles: &TestPrecompiles<Runtime>,
+	action: Action,
+) -> PrecompilesTester<TestPrecompiles<Runtime>> {
+	let (_, gas_reserve) = costs(action);
+
+	precompiles
+		.prepare_test(
+			Alice,
+			Precompile,
+			EvmDataWriter::new_with_selector(action)
+				.write(vec![Address(Bob.into())])
+				.write(vec![U256::from(1u8)])
+				.write(vec![Bytes::from(b"one".as_slice())])
+				.write::<Vec<U256>>(vec![])
+				.write(true)
+				.build(),
+		)
+		.with_target_gas(Some(50_000))
+		.with_subcall_handle(move |subcall| {
+			let Subcall {
+				address,
+				transfer,
+				input,
+				target_gas,
+				is_static,
+				context,
+			} = subcall;
+
+			// Called from the precompile caller.
+			assert_eq!(context.caller, Alice.into());
+			assert_eq!(is_static, false);
+
+			match address {
+				a if a == Bob.into() => {
+					assert_eq!(
+						target_gas,
+						Some(50_000 - gas_reserve),
+						"batch forward all gas"
+					);
+					let transfer = transfer.expect("there is a transfer");
+					assert_eq!(transfer.source, Alice.into());
+					assert_eq!(transfer.target, Bob.into());
+					assert_eq!(transfer.value, 1u8.into());
+
+					assert_eq!(context.address, Bob.into());
+					assert_eq!(context.apparent_value, 1u8.into());
+
+					assert_eq!(&input, b"one");
+
+					SubcallOutput {
+						reason: ExitReason::Error(ExitError::OutOfGas),
+						output: Vec::new(),
+						cost: 11_000,
+						logs: vec![],
+					}
+				}
+				_ => panic!("unexpected subcall"),
+			}
+		})
+}
+
+#[test]
+fn batch_some_out_of_gas() {
+	ExtBuilder::default().build().execute_with(|| {
+		batch_out_of_gas(&precompiles(), Action::BatchSome)
+			.expect_log(log_subcall_failed(Precompile, 0))
+			.execute_returns(Vec::new())
+	})
+}
+
+#[test]
+fn batch_some_until_failure_out_of_gas() {
+	ExtBuilder::default().build().execute_with(|| {
+		batch_out_of_gas(&precompiles(), Action::BatchSomeUntilFailure)
+			.expect_log(log_subcall_failed(Precompile, 0))
+			.execute_returns(Vec::new())
+	})
+}
+
+#[test]
+fn batch_all_out_of_gas() {
+	ExtBuilder::default().build().execute_with(|| {
+		batch_out_of_gas(&precompiles(), Action::BatchAll).execute_error(ExitError::OutOfGas)
+	})
+}
```

Comment:
- batchAll : only succeed if all call succeed, otherwise forward errors as soon as they occur (thus propagating OOG).
- batchSome* : the batch call itself always succeed, but subcalls may fail. If a subcall fails batchSome will continue with following calls, while batchSomeUntilFailure will stop there (while not reverting).